### PR TITLE
Automate plotting

### LINKED ISSUE #33: Add plotting to analysis pipeline
The plots added in #30 should be generated automatically for each run of the analysis pipeline.

### PR DESCRIPTION
**Pull request by mcwitt**
Closes #33, #48 

#### Work distributions for solvent and complex (`RUN0`)
RUN0

#### Convergence of relative free energies by GEN (`RUN0-convergence`)
RUN0-convergence

#### Distribution of relative binding free energies (`relative_fe_dist`)
relative_fe_dist

#### Cumulative distribution of relative free binding energies (`cumulative_fe_dist`)
cumulative_fe_hist

#### TODO
- [x] Add convergence plot (requires #36)
- [x] Add missing docstrings

### CODE REVIEW

File: covid_moonshot/analysis/fah_plotting.py [hunk 1]
```diff
@@ -76,49 +60,73 @@ def plot_relative_distribution(relative_fes, bins=100, title='Relative affinity'
...
-        alpha=0.5,
-        label=f"N={len(relative_fes)}",
+        relative_fes, color="hotpink", alpha=0.5, label=f"N={len(relative_fes)}",
     )
     plt.xlabel("Relative free energy to ligand 0 / kcal/mol")
     plt.title(title)
-    _produce_plot(filename)
 
 
 def plot_convergence(results, n_gens=3, title=None):
```
Comment (mcwitt) [inline covid_moonshot/analysis/fah_plotting.py:75]: @hannahbrucemacdonald @jchodera it looks like this plot relies on results at the GEN level, which we are not yet computing in the analysis pipeline. Is it worth adding this analysis, or better to skip this plot for now?

Comment (jchodera) [reply covid_moonshot/analysis/fah_plotting.py:75]: Let's break out the GEN-level info into a separate issue to first discuss how to represent work values can be organized by GEN. That's something that will be important as a sanity check.

**Comment by jchodera** [comment]
@mcwitt Can you summarize what gets plotted (with examples) and where it ends up?

File: covid_moonshot/analysis/fah_plotting.py [hunk 2]
```diff
@@ -1,154 +1,191 @@
+import matplotlib.figure
```
Comment (jchodera) [inline covid_moonshot/analysis/fah_plotting.py:1]: Not necessary to do this now, but in general, we recommend lazy imports inside methods that use external packages so that if we reuse components of this in other tools, (1) it doesn't take forever to import, and (2) we can have optional dependencies for certain functionality. This also helps enormously when using readthedocs, since we don't need to have all dependencies installed just for sphinx to inspect the docstrings and generate API docs.

File: covid_moonshot/analysis/fah_plotting.py [hunk 3]
```diff
@@ -1,154 +1,191 @@
...
 
-def _produce_plot(name, show_plots=False, plot_file_format="pdf"):
-    if show_plots:
-        plt.show()
-    else:
-        fname = os.extsep.join([name, plot_file_format])
-        logging.info(f"Writing {fname}")
-        plt.savefig(fname)
-
+def plot_single_work_distribution(
```
Comment (jchodera) [inline covid_moonshot/analysis/fah_plotting.py:16]: Missing docstring?

File: covid_moonshot/analysis/fah_plotting.py [hunk 4]
```diff
@@ -1,154 +1,191 @@
...
     )
-    if title is not None:
-        ax.set_title(title)
 
 
-def plot_two_work_distribution(f1, r1, f2, r2, phases=(None, None), title=None):
-    fig, (ax1, ax2) = plt.subplots(ncols=2, nrows=1, figsize=(7.5, 3.25))
-    plot_single_work_distribution(f1, r1, ax1, phases[0])
-    plot_single_work_distribution(f2, r2, ax2, phases[1])
+def plot_two_work_distribution(
```
Comment (jchodera) [inline covid_moonshot/analysis/fah_plotting.py:39]: Missing docstring?

File: covid_moonshot/analysis/fah_plotting.py [hunk 5]
```diff
@@ -1,154 +1,191 @@
...
-        plt.legend()
-        _produce_plot(f"fe_convergence_{title}")
-
-
-def plot_cumulative_distributions(results, minimum=None, maximum=5, cmap='PiYG', n_bins=100,
-                                  markers=[-2, -1, 0, 1, 2], title='Cumulative distribution',
-                                  filename='Cumulative distribution'):
+
+
+def plot_convergence(
```
Comment (jchodera) [inline covid_moonshot/analysis/fah_plotting.py:102]: Missing docstring?

File: covid_moonshot/analysis/fah_plotting.py [hunk 6]
```diff
@@ -1,154 +1,191 @@
...
+    gens: List[int],
+    complex_delta_fs: List[float],
+    complex_delta_f_errs: List[float],
```
Comment (jchodera) [inline covid_moonshot/analysis/fah_plotting.py:105]: Should we move toward a data structure that encapsulates the phase (complex, solvent) and overall (binding) free energies and uncertainties in a single object?

File: covid_moonshot/analysis/fah_plotting.py [hunk 7]
```diff
@@ -179,19 +216,83 @@ def plot_cumulative_distributions(results, minimum=None, maximum=5, cmap='PiYG',
...
+            f"$N$ = {len([x for x in results if x < v])}",
+            rotation=90,
+            verticalalignment="center",
+            color="green",
+        )
+    plt.xlabel("Affinity relative to ligand 0 / " + r"kcal mol$^{-1}$")
+    plt.ylabel("Cumulative $N$ ligands")
+
+
+def get_plot_filename(path: str, name: str, file_format: str) -> str:
```
Comment (jchodera) [inline covid_moonshot/analysis/fah_plotting.py:238]: Missing docstring?

File: covid_moonshot/analysis/fah_plotting.py [hunk 8]
```diff
@@ -179,19 +216,83 @@ def plot_cumulative_distributions(results, minimum=None, maximum=5, cmap='PiYG',
...
+    return os.path.join(path, os.extsep.join([name, file_format]))
+
+
+def save_run_level_plots(
```
Comment (jchodera) [inline covid_moonshot/analysis/fah_plotting.py:242]: Missing docstring?

File: covid_moonshot/analysis/fah_plotting.py [hunk 9]
```diff
@@ -179,19 +216,83 @@ def plot_cumulative_distributions(results, minimum=None, maximum=5, cmap='PiYG',
...
+        complex_delta_f_errs=[gen.free_energy.ddelta_f for gen in complex_phase.gens],
+        solvent_delta_fs=[gen.free_energy.delta_f for gen in solvent_phase.gens],
+        solvent_delta_f_errs=[gen.free_energy.ddelta_f for gen in solvent_phase.gens],
+        binding_delta_f=binding.delta_f,
+        binding_delta_f_err=binding.ddelta_f,
+    )
+    plt.savefig(get_plot_filename(path, f"run{run}-convergence", file_format))
+
+
+def save_summary_plots(
```
Comment (jchodera) [inline covid_moonshot/analysis/fah_plotting.py:285]: Missing docstring?

File: covid_moonshot/lib.py [hunk 10]
```diff
@@ -123,6 +125,14 @@ def analyze_run(
         except ValueError as e:
             raise ValueError(f"Failed to save structures for complex: {e}")
 
+    save_run_level_plots(
```
Comment (jchodera) [inline covid_moonshot/lib.py:128]: Missing docstring?

File: setup.py [hunk 11]
```diff
@@ -12,7 +12,9 @@
         "joblib>=0.16",
```
Comment (jchodera) [inline setup.py:12]: Should we also have an `environment.yml` to make it easier to configure the run environment? The new MolSSI Cookiecutter requires this, so we can wait until we migrate.

File: covid_moonshot/analysis/fah_plotting.py (hunk 3 above)
Comment (mcwitt) [reply covid_moonshot/analysis/fah_plotting.py]: Added now (mostly 74c4814)

File: covid_moonshot/analysis/fah_plotting.py (hunk 4 above)
Comment (mcwitt) [reply covid_moonshot/analysis/fah_plotting.py]: 74c4814

File: covid_moonshot/analysis/fah_plotting.py (hunk 5 above)
Comment (mcwitt) [reply covid_moonshot/analysis/fah_plotting.py]: 74c4814

File: covid_moonshot/analysis/fah_plotting.py (hunk 6 above)
Comment (mcwitt) [reply covid_moonshot/analysis/fah_plotting.py]: Yes, this already exists as a matter of fact (the `RunAnalysis` object). I intentionally kept the interface to the plotting functions low-level (i.e. not using the types in `covid_moonshot.core`) in case they would be useful to call from other code where we don't want the full baggage of the object hierarchy. I could go either way though, honestly.

File: covid_moonshot/analysis/fah_plotting.py (hunk 2 above)
Comment (mcwitt) [reply covid_moonshot/analysis/fah_plotting.py]: Makes sense. I'd been doing this mostly with modules that aren't in PyPI / in `install_requires`, but I can be more aggressive about this.

File: covid_moonshot/lib.py (hunk 10 above)
Comment (mcwitt) [reply covid_moonshot/lib.py:128]: Addressed by 41e9e6a, 6717294

File: covid_moonshot/analysis/fah_plotting.py (hunk 9 above)
Comment (mcwitt) [reply covid_moonshot/analysis/fah_plotting.py]: Added in 41e9e6a

File: covid_moonshot/analysis/fah_plotting.py (hunk 7 above)
Comment (mcwitt) [reply covid_moonshot/analysis/fah_plotting.py]: Addressed by 41e9e6a, 6717294

File: covid_moonshot/lib.py (hunk 10 above)
Comment (mcwitt) [reply covid_moonshot/lib.py:128]: Added in 74c4814

File: covid_moonshot/analysis/fah_plotting.py (hunk 8 above)
Comment (mcwitt) [reply covid_moonshot/analysis/fah_plotting.py]: Added in 74c4814

File: setup.py (hunk 11 above)
Comment (mcwitt) [reply setup.py:12]: Definitely on board with this! I'm having an issue with conda currently (my version doesn't have the `--from-history` option for `conda env export`, and the newer versions seem broken for my OS). I'll plan to add it to a later PR, or when we do the migration.

**Comment by mcwitt** [comment]
@jchodera I added a brief summary of the run-level and summary plots to the docstrings of the relevant functions in 68c787b. Also added example images to the PR description.

**Comment by jchodera** [comment]
This is amazing!

For the work distributions:
* Can we use "$k_B T$ instead of having "300 K" in there?
* Let's add a vertical black solid or dashed line for the free energy estimate from all runs, if it's easy to do that. It should appear near the intersection of the normalized work distributions, but is a useful visual check/guide.

Otherwise, this looks _fantastic_!

File: covid_moonshot/analysis/plots.py [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+    ax: plt.Axes, forward_works: List[float], reverse_works: List[float]
+) -> None:
+    """
+    Plot a single work distribution
+
+    Parameters
+    ----------
+    ax : AxesSubplot
+       Axes on which to draw the plot
+    forward_works : list of float
```
Comment (jchodera) [inline covid_moonshot/analysis/plots.py:27]: Apologies for just realizing this: Would be good to document units for work or energy values if they are floats! You can just say "(in kT)" for all these.

**Comment by mcwitt** [comment]
@jchodera 
> Let's add a vertical black solid or dashed line for the free energy estimate from all runs
I'm a bit confused here. Do you mean the binding free energy estimate for the current run? (We're not calculating any aggregates over all runs in the pipeline)

File: covid_moonshot/analysis/plots.py (hunk 12 above)
Comment (mcwitt) [reply covid_moonshot/analysis/plots.py:27]: Cool, I can add this. My convention so far is to assume kT and specify only when we differ from that, but explicit is better than implicit!

Comment (mcwitt) [reply covid_moonshot/analysis/plots.py:27]: Will be cool to transition to `pint` or something similar!

**Comment by jchodera** [comment]
Whoops! Each phase (complex, solvent) has a BAR free energy estimate, and when plotted as a vertical line at the free energy estimate on those work distribution plots, should appear near the intersection of distributions. I can draw a sketch if that would be helpful!

**Comment by mcwitt** [comment]
@jchodera makes sense! Added in 8493d31 and updated the example image.

File: covid_moonshot/analysis/plots.py (hunk 12 above)
Comment (mcwitt) [reply covid_moonshot/analysis/plots.py:30]: Annotated docstrings with `(in kT)` for now: 7b59b1e

**Comment by jchodera** [comment]
💯

File: covid_moonshot/analysis/plots.py [hunk 13]
```diff
@@ -0,0 +1,448 @@
...
+    cmap: str = "PiYG",
+    n_bins: int = 100,
+    markers: List[float] = [-2, -1, 0, 1, 2],
+) -> None:
+    """
+    Plot cumulative distribution of ligand affinities
+
+    Parameters
+    ----------
+    affinities : list of float
```
Comment (jchodera) [inline covid_moonshot/analysis/plots.py:267]: What units are these supposed to be in?
Let's also be a bit careful to call then "delta_f" or "free_energies" since the relationship between the free energy and affinity is
```
affinity = exp(-free_energy / kT) * (1 Molar)
```
We generally don't want to use affinities in our reports for internal use.

**Comment by jchodera** [comment]
I've activated this branch on `aws3` to test the plotting pipeline, and there appears to be a discrepancy between the PDF/CDF plots and the summarized `ligands.pdf`. The CDF shows the best ligands are < -2 kcal/mol:
image
But the `ligands.pdf` file the `postprocess.py` script (which assumes the JSON outputs binding free energies in kT and converts to kcal/mol) generates shows no ligands < -2 kcal/mol:
image
We should be a bit careful with units to use kT for all serialization but we can plot with kcal/mol.

File: covid_moonshot/analysis/plots.py (hunk 13 above)
Comment (mcwitt) [reply covid_moonshot/analysis/plots.py]: Ah! I wasn't aware of this convention. Makes sense. I replaced all occurrences of "affinity" with "delta_f" or "free energy" in 9eddd3d, 56cfc0b

**Comment by mcwitt** [comment]
> I've activated this branch on aws3 to test the plotting pipeline, and there appears to be a discrepancy between the PDF/CDF plots 

@jchodera I think this is due to the comparison of values in kT and kcal/mol, fixed in 54a517d. I'll try to reproduce your example to make sure.

**Comment by mcwitt** [comment]
> We should be a bit careful with units to use kT for all serialization but we can plot with kcal/mol.

The convention I'm using now is:
- All free energies are in `kT`, unless the name is explicitly suffixed with `_kcal`
- Values in kcal/mol only exist in the plotting code `analysis.plots` (everything we serialize is kT)

I'll look at adding `pint` or something similar soon. It seems really useful to be able to catch errors like the one fixed in 
54a517d at runtime (or even better, with a red underline in the editor, if there is mypy support)

**Comment by jchodera** [comment]
Whoops, not sure why these commits showed up as being from @hannahbrucemacdonald, but 
17afc59 should fix the off-by-one problem (#46)!

**Comment by hannahbrucemacdonald** [comment]
Looks good to me! Any reason that the convergence plots are deltaF rather than deltaG?

**Review by hannahbrucemacdonald — APPROVED** [review]

**Comment by mcwitt** [comment]
> Looks good to me! Any reason that the convergence plots are deltaF rather than deltaG?

No good reason! Delta G is more appropriate, correct? I can change it.
